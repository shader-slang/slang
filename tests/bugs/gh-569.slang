// gh-569.slang
//TEST(compute):COMPARE_COMPUTE: -shaderobj

// Test that correct scoping is used in generated HLSL/GLSL,
// even when dominator tree and structured control flow disagree.

uint test(uint inVal)
{
	uint tmp = inVal;
    for(;;)
    {
    	if(tmp < 4)
    	{
    		tmp++;
    	}
    	else
    	{
    		break;
    	}
    }
    return tmp + inVal;
}

//TEST_INPUT:ubuffer(data=[0 0 0 0], stride=4):out,name gBuffer
RWStructuredBuffer<uint> gBuffer;

[numthreads(4, 1, 1)]
void computeMain(uint3 dispatchThreadID : SV_DispatchThreadID)
{
    uint tid = dispatchThreadID.x;

    uint val = tid;
    val = test(val);

    gBuffer[tid] = val;
}
