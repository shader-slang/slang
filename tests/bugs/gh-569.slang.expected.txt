4
5
6
7
